### PR TITLE
Enable nullability in several internal CodeDomSerializers

### PR DESCRIPTION
**Pull request by halgab**
## Proposed changes

- Enable nullability in `ContainerCodeDomSerializer`
- Enable nullability in `EnumCodeDomSerializer`
- Enable nullability in `EventMemberCodeDomSerializer`
- Enable nullability in `PrimitiveCodeDomSerializer`
- Enable nullability in `ResourceCodeDomSerializer`
- Enable nullability in `ToolStripMenuItemCodeDomSerializer`
- Enable nullability in `ComponentCodeDomSerializer` (2nd commit)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9842)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ComponentCodeDomSerializer.cs [hunk 1]
```diff
@@ -278,50 +258,47 @@ public override object Serialize(IDesignerSerializationManager manager, object v
...
                         // the component wants a special IContainer constructor or not.  For that to be valid we must also know
                         // that we can get to an actual IContainer.
-                        IContainer container = manager.GetService(typeof(IContainer)) as IContainer;
-                        ConstructorInfo ctor = null;
+                        IContainer? container = manager.GetService<IContainer>();
                         if (container is not null)
                         {
-                            ctor = GetReflectionTypeHelper(manager, value).GetConstructor(BindingFlags.ExactBinding | BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly, null, GetContainerConstructor(manager), null);
-                        }
+                            ConstructorInfo? ctor = GetReflectionTypeHelper(manager, value).GetConstructor(BindingFlags.ExactBinding | BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly, null, GetContainerConstructor(manager), null);
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ComponentCodeDomSerializer.cs:264]: nit
```suggestion
                            ConstructorInfo? ctor = GetReflectionTypeHelper(manager, value).GetConstructor(
                            BindingFlags.ExactBinding | BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly, 
                            null, 
                            GetContainerConstructor(manager), 
                            null);
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ComponentCodeDomSerializer.cs [hunk 2]
```diff
@@ -416,15 +393,19 @@ public override object Serialize(IDesignerSerializationManager manager, object v
...
                                     foreach (object dependency in oldEntry.Dependencies)
                                     {
                                         entry.AddDependency(dependency);
                                     }
                                 }
                             }
+                            else
+                            {
+                                entry.Statements = new CodeStatementCollection();
+                            }
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ComponentCodeDomSerializer.cs:408]: Why was this added?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/EnumCodeDomSerializer.cs [hunk 3]
```diff
@@ -12,52 +10,44 @@ namespace System.ComponentModel.Design.Serialization;
...
         {
             Trace(TraceLevel.Verbose, $"Type: {(value is null ? "(null)" : value.GetType().Name)}");
-            if (value is Enum)
+            if (value is Enum enumValue)
             {
                 bool needCast = false;
                 Enum[] values;
-                TypeConverter converter = TypeDescriptor.GetConverter(value);
+                TypeConverter converter = TypeDescriptor.GetConverter(enumValue);
                 if (converter is not null && converter.CanConvertTo(typeof(Enum[])))
```
[the remark this answers is not in the document]
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/EnumCodeDomSerializer.cs:35]: Nullable annotations say it can't happen, but looking at the upstream code, it could happen in Release (there is an assert). I'll update

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ComponentCodeDomSerializer.cs (hunk 2 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ComponentCodeDomSerializer.cs:408]: To have `ComponentCache.Entry.Statements`  declared as not null, I had to create the `Statements` collection with the `Entry` class. So to avoid throwing away an empty collection we've just created, I swapped old line 447 with this code, so that we create a new collection only if we are reusing `entry`

**Review by halgab — COMMENTED** [review]
I still have to figure out why the tests are failing... I have had trouble running them locally since the rebranding though

**Comment by lonitra** [comment]
> I have had trouble running them locally since the rebranding though

😮 Are you getting errors when running test locally?

**Comment by halgab** [comment]
The `start-vs` script stopped working (main branch, current head), so I have to open VS (17.8.0 preview 1) manually:

image

Then, when I try to launch the tests using the test explorer, I get the following output from Tests:

```
System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Microsoft.VisualStudio.TestWindow.Controller.TestContainerConfigurationQueryByTestsBase.<GetTestContainerConfigurations>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.Controller.TestContainerConfigurationQueryByTestsBase.<GetContainersToUpdateInternalAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.Controller.TestContainerConfigurationQuery.<GetAndCacheContainersToUpdateAsync>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.Controller.RunOperation.<FindChangedContainersAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.Controller.DiscoveryOperation.<DetectChangedContainersAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.Controller.RunOperation.<RunTestsAsync>d__22.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.TestWindow.Controller.RunOperation.<RunTestsAsync>d__22.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.Controller.RunOperation.<ExecuteInternalAsync>d__23.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.Controller.Operation.<<ExecuteAsync>b__43_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.TestWindow.Logging.ILoggerExtensions.<CallWithCatchAsync>d__11`1.MoveNext()
```

**Comment by halgab** [comment]
I still managed to find the bug without debugging the tests. This is ready now

**Review by lonitra — APPROVED** [review]
LGTM 👍
